extensions/ql-vscode/src/remote-queries/repository-selection.ts

```diff
@@ -1,54 +1,116 @@
+    ...createSystemDefinedRepoListsQuickPickItems(),
+    ...createUserDefinedRepoListsQuickPickItems(),
+  ];
+
+  const options = {
+    placeHolder: 'Select a repository list. You can define repository lists in the `codeQL.variantAnalysis.repositoryLists` setting.',
+    ignoreFocusOut: true,
+  };
+
+  const quickpick = await window.showQuickPick<RepoListQuickPickItem>(
+    quickPickItems,
+    options);
+
+  if (quickpick?.repositories?.length) {
+    void logger.log(`Selected repositories: ${quickpick.repositories.join(', ')}`);
+    return { repositories: quickpick.repositories };
+  } else if (quickpick?.repositoryList) {
+    void logger.log(`Selected repository list: ${quickpick.repositoryList}`);
+    return { repositoryLists: [quickpick.repositoryList] };
+  } else if (quickpick?.useCustomRepository) {
+    const customRepo = await getCustomRepo();
+    if (!customRepo || !REPO_REGEX.test(customRepo)) {
+      void showAndLogErrorMessage('Invalid repository format. Please enter a valid repository in the format <owner>/<repo> (e.g. github/codeql)');
+      return {};
     }
+    void logger.log(`Entered repository: ${customRepo}`);
+    return { repositories: [customRepo] };
   } else {
-    void logger.log('No repository lists defined. Displaying text input box.');
-    const remoteRepo = await window.showInputBox({
-      title: 'Enter a GitHub repository in the format <owner>/<repo> (e.g. github/codeql)',
-      placeHolder: '<owner>/<repo>',
-      prompt: 'Tip: you can save frequently used repositories in the `codeQL.variantAnalysis.repositoryLists` setting',
-      ignoreFocusOut: true,
-    });
-    if (!remoteRepo) {
-      void showAndLogErrorMessage('No repositories entered.');
-      return;
-    } else if (!REPO_REGEX.test(remoteRepo)) { // Check if user entered invalid input
-      void showAndLogErrorMessage('Invalid repository format. Must be in the format <owner>/<repo> (e.g. github/codeql)');
-      return;
-    }
-    void logger.log(`Entered repository: ${remoteRepo}`);
-    return [remoteRepo];
+    void showAndLogErrorMessage('No repositories selected.');
+    return {};
+  }
+}
+
+/**
+ * Checks if the selection is valid or not.
+ * @param repoSelection The selection to check.
+ * @returns A boolean flag indicating if the selection is valid or not.
+ */
+export function isInvalidSelection(repoSelection: RepositorySelection): boolean {
+  if (repoSelection.repositories === undefined && repoSelection.repositoryLists === undefined) {
+    return true;
+  }
+  if (repoSelection.repositories !== undefined && repoSelection.repositories.length === 0) {
+    return true;
   }
+  if (repoSelection.repositoryLists !== undefined && repoSelection.repositoryLists.length === 0) {
+    return true;
+  }
+
+  return false;
+}
+
+function createSystemDefinedRepoListsQuickPickItems(): RepoListQuickPickItem[] {
+  const topNs = [10, 100, 1000];
+
+  return topNs.map(n => ({
+    label: '$(star) Top ' + n,
```

Comment:
Yes. I think that would be a bad user experience. All interactions should preferably be at the beginning. 

It would certainly be possible to determine the language in a separate step before choosing the repo and building the pack, but that refactoring is a bigger change and not worth it right now in my opinion. Let's just keep it as it is and we can rethink later.